cellarium/ml/layers/attention.py

```diff
@@ -0,0 +1,164 @@
+            create_initializer(self.Wqkv_initializer)(module.weight)
+            if module.bias is not None:
+                nn.init.zeros_(module.bias)
+
+        create_initializer(self.Wo_initializer)(self.Wo.weight)
+        if self.Wo.bias is not None:
+            nn.init.zeros_(self.Wo.bias)
+
+    @staticmethod
+    def split_heads(X_nqd: torch.Tensor, n_heads: int) -> torch.Tensor:
+        """Transposition for parallel computation of multiple attention heads."""
+        X_nqhk = X_nqd.reshape(X_nqd.shape[0], X_nqd.shape[1], n_heads, -1)
+        X_nhqk = X_nqhk.permute(0, 2, 1, 3)
+        return X_nhqk
+
+    @staticmethod
+    def merge_heads(X_nhqk: torch.Tensor) -> torch.Tensor:
+        """Reverse of split_heads."""
+        X_nqhk = X_nhqk.permute(0, 2, 1, 3)
+        X_nqd = X_nqhk.reshape(X_nqhk.shape[0], X_nqhk.shape[1], -1)
+        return X_nqd
+
+    def forward(
+        self,
+        query_ncd: torch.Tensor,
+        key_ncd: torch.Tensor,
+        value_ncd: torch.Tensor,
+        attention_mask_ncc: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Args:
+            query_ncd:
+                Query tensor of shape ``(n, c, d)``.
+            key_ncd:
+                Key tensor of shape ``(n, c, d)``.
+            value_ncd:
+                Value tensor of shape ``(n, c, d)``.
+            attention_mask_ncc:
+                Attention mask tensor of shape ``(n, c, c)``.
+
+        Returns:
+            The output hidden state tensor of shape ``(n, c, d)``.
+        """
+        n_heads = self.n_heads
+        query_ncd = self.Wq(query_ncd)
+        key_ncd = self.Wk(key_ncd)
+        value_ncd = self.Wv(value_ncd)
+        # d = k * h
+        query_nhck = self.split_heads(query_ncd, n_heads)
+        key_nhck = self.split_heads(key_ncd, n_heads)
+        value_nhck = self.split_heads(value_ncd, n_heads)
+
+        # scale_factor is computed according to the muP paper
+        scale_factor = self.attention_logits_scale / query_nhck.shape[-1]
+
+        if (self.attention_backend == "torch") or use_cs():
+            key_nhck = key_nhck * torch.tensor(scale_factor, dtype=key_nhck.dtype)
+            attention_logits_nhcc = torch.matmul(query_nhck, key_nhck.transpose(-1, -2))
+            neg_inf = torch.tensor(float("-inf"), dtype=torch.float32)
+            attention_bias_ncc = torch.where(attention_mask_ncc, 0, neg_inf).to(attention_logits_nhcc.dtype)
+            attention_logits_nhcc += attention_bias_ncc.unsqueeze(1).expand(attention_logits_nhcc.shape)
+            if self.attention_softmax_fp32 and attention_logits_nhcc.dtype != torch.float32:
+                attention_weights_nhcc = nn.functional.softmax(attention_logits_nhcc.float(), dim=-1).to(
+                    attention_logits_nhcc.dtype
+                )
+            else:
+                attention_weights_nhcc = nn.functional.softmax(attention_logits_nhcc, dim=-1)
+            attention_weights_nhcc = nn.functional.dropout(
```

Comment:
Is it common to do dropout after softmax, or before (e.g. setting some random keys to -inf)?

Reply:
PyTorch implementation does it after the softmax - https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html